File: backend/app/core/monitoring.py
"""
Monitoring module for Sentry and Prometheus integration.
"""
import os
import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from prometheus_client import Counter, Histogram, Gauge, generate_latest, CONTENT_TYPE_LATEST
from typing import Optional

# Prometheus metrics
REQUEST_COUNT = Counter(
    'http_requests_total',
    'Total HTTP requests',
    ['method', 'endpoint', 'status']
)

REQUEST_LATENCY = Histogram(
    'http_request_duration_seconds',
    'HTTP request latency',
    ['method', 'endpoint']
)

SEARCH_COUNT = Counter(
    'search_requests_total',
    'Total search requests by country',
    ['country', 'status']
)

CACHE_HITS = Counter(
    'cache_hits_total',
    'Total cache hits'
)

CACHE_MISSES = Counter(
    'cache_misses_total',
    'Total cache misses'
)

ACTIVE_CONNECTIONS = Gauge(
    'active_connections',
    'Number of active connections'
)

DB_QUERY_COUNT = Counter(
    'db_queries_total',
    'Total database queries',
    ['operation']
)

DB_QUERY_LATENCY = Histogram(
    'db_query_duration_seconds',
    'Database query latency',
    ['operation']
)


def init_sentry(dsn: Optional[str] = None, traces_sample_rate: float = 0.1, environment: str = "development"):
    """
    Initialize Sentry SDK for error tracking.
    
    Args:
        dsn: Sentry DSN (Data Source Name)
        traces_sample_rate: Percentage of traces to send to Sentry (0.0 to 1.0)
        environment: Environment name (development, staging, production)
    """
    if not dsn:
        return
    
    sentry_sdk.init(
        dsn=dsn,
        traces_sample_rate=traces_sample_rate,
        environment=environment,
        integrations=[
            FastApiIntegration(),
            SqlalchemyIntegration(),
        ],
        # Filter sensitive data
        before_send=filter_sensitive_data,
    )


def filter_sensitive_data(event, hint):
    """
    Filter sensitive data from Sentry events.
    
    Removes passwords, tokens, and other sensitive information.
    """
    # Remove sensitive headers
    if 'request' in event and 'headers' in event['request']:
        headers = event['request']['headers']
        sensitive_headers = ['authorization', 'cookie', 'x-api-key']
        for header in sensitive_headers:
            if header in headers:
                headers[header] = '[Filtered]'
    
    # Remove sensitive POST data
    if 'request' in event and 'data' in event['request']:
        data = event['request']['data']
        if isinstance(data, dict):
            sensitive_fields = ['password', 'token', 'secret', 'api_key']
            for field in sensitive_fields:
                if field in data:
                    data[field] = '[Filtered]'
    
    return event


def capture_exception(exception: Exception, context: Optional[dict] = None):
    """
    Capture an exception and send it to Sentry.
    
    Args:
        exception: The exception to capture
        context: Additional context to include
    """
    with sentry_sdk.push_scope() as scope:
        if context:
            for key, value in context.items():
                scope.set_context(key, value)
        sentry_sdk.capture_exception(exception)


def capture_message(message: str, level: str = "info", context: Optional[dict] = None):
    """
    Capture a message and send it to Sentry.
    
    Args:
        message: The message to capture
        level: Message level (debug, info, warning, error, fatal)
        context: Additional context to include
    """
    with sentry_sdk.push_scope() as scope:
        if context:
            for key, value in context.items():
                scope.set_context(key, value)
        sentry_sdk.capture_message(message, level=level)


def get_metrics():
    """
    Get Prometheus metrics in the format expected by Prometheus.
    
    Returns:
        Tuple of (metrics_data, content_type)
    """
    return generate_latest(), CONTENT_TYPE_LATEST

File: frontend/src/utils/monitoring.js
/**
 * Frontend monitoring with Sentry integration.
 */
import * as Sentry from '@sentry/react';

/**
 * Initialize Sentry SDK for error tracking.
 * 
 * @param {Object} options - Configuration options
 * @param {string} options.dsn - Sentry DSN
 * @param {number} options.tracesSampleRate - Percentage of traces to send (0.0 to 1.0)
 * @param {string} options.environment - Environment name
 */
export function initSentry({ 
  dsn = import.meta.env.VITE_SENTRY_DSN,
  tracesSampleRate = parseFloat(import.meta.env.VITE_SENTRY_TRACES_SAMPLE_RATE || '0.1'),
  environment = import.meta.env.VITE_ENVIRONMENT || 'development'
} = {}) {
  if (!dsn) {
    console.log('Sentry DSN not configured, skipping initialization');
    return;
  }

  Sentry.init({
    dsn,
    integrations: [
      Sentry.browserTracingIntegration(),
      Sentry.replayIntegration({
        maskAllText: true,
        blockAllMedia: true,
      }),
    ],
    
    // Performance Monitoring
    tracesSampleRate,
    
    // Session Replay
    replaysSessionSampleRate: 0.1, // Sample 10% of sessions
    replaysOnErrorSampleRate: 1.0, // Sample 100% of sessions with errors
    
    environment,
    
    // Filter sensitive data
    beforeSend(event, hint) {
      // Remove sensitive data from URLs
      if (event.request && event.request.url) {
        event.request.url = event.request.url.replace(/([?&])(token|password|api_key)=[^&]*/gi, '$1$2=[Filtered]');
      }
      
      // Remove sensitive data from breadcrumbs
      if (event.breadcrumbs) {
        event.breadcrumbs = event.breadcrumbs.map(breadcrumb => {
          if (breadcrumb.data) {
            const filtered = { ...breadcrumb.data };
            ['password', 'token', 'api_key', 'secret'].forEach(key => {
              if (key in filtered) {
                filtered[key] = '[Filtered]';
              }
            });
            return { ...breadcrumb, data: filtered };
          }
          return breadcrumb;
        });
      }
      
      return event;
    },
  });

  console.log('Sentry initialized successfully');
}

/**
 * Capture an exception and send it to Sentry.
 * 
 * @param {Error} error - The error to capture
 * @param {Object} context - Additional context
 */
export function captureException(error, context = {}) {
  Sentry.withScope((scope) => {
    // Add context
    Object.keys(context).forEach(key => {
      scope.setContext(key, context[key]);
    });
    
    // Capture exception
    Sentry.captureException(error);
  });
}

/**
 * Capture a message and send it to Sentry.
 * 
 * @param {string} message - The message to capture
 * @param {string} level - Message level (debug, info, warning, error, fatal)
 * @param {Object} context - Additional context
 */
export function captureMessage(message, level = 'info', context = {}) {
  Sentry.withScope((scope) => {
    // Add context
    Object.keys(context).forEach(key => {
      scope.setContext(key, context[key]);
    });
    
    // Capture message
    Sentry.captureMessage(message, level);
  });
}

/**
 * Set user context for Sentry events.
 * 
 * @param {Object} user - User information
 * @param {string} user.id - User ID
 * @param {string} user.email - User email
 * @param {string} user.username - Username
 */
export function setUser(user) {
  if (user) {
    Sentry.setUser({
      id: user.id,
      email: user.email,
      username: user.username,
    });
  } else {
    // Clear user context
    Sentry.setUser(null);
  }
}

/**
 * Add breadcrumb for debugging.
 * 
 * @param {Object} breadcrumb - Breadcrumb data
 * @param {string} breadcrumb.message - Breadcrumb message
 * @param {string} breadcrumb.category - Breadcrumb category
 * @param {string} breadcrumb.level - Breadcrumb level
 * @param {Object} breadcrumb.data - Additional data
 */
export function addBreadcrumb({ message, category = 'custom', level = 'info', data = {} }) {
  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data,
  });
}

export default {
  initSentry,
  captureException,
  captureMessage,
  setUser,
  addBreadcrumb,
};
